app/models/reviewable.rb

```diff
@@ -0,0 +1,395 @@
+  end
+
+  def self.count_by_date(start_date, end_date, category_id = nil)
+    scores_with_topics
+      .where('reviewable_scores.created_at BETWEEN ? AND ?', start_date, end_date)
+      .where("topics.category_id = COALESCE(?, topics.category_id)", category_id)
+      .group("date(reviewable_scores.created_at)")
+      .order('date(reviewable_scores.created_at)')
+      .count
+  end
+
+protected
+
+  def recalculate_score
+    # Recalculate the pending score and return it
+    result = DB.query(<<~SQL, id: self.id, pending: ReviewableScore.statuses[:pending])
+      UPDATE reviewables
+      SET score = COALESCE((
+        SELECT sum(score)
+        FROM reviewable_scores AS rs
+        WHERE rs.reviewable_id = :id
+          AND rs.status = :pending
+      ), 0.0)
+      WHERE id = :id
+      RETURNING score
+    SQL
+
+    # Update topic score
+    DB.query(<<~SQL, topic_id: topic_id, pending: Reviewable.statuses[:pending])
+      UPDATE topics
+      SET reviewable_score = COALESCE((
+        SELECT SUM(score)
+        FROM reviewables AS r
+        WHERE r.topic_id = :topic_id AND r.status = :pending
+      ), 0.0)
+      WHERE id = :topic_id
+    SQL
+
+    self.score = result[0].score
+  end
+
+  def increment_version!(version = nil)
+    version_result = nil
+
+    if version
+      version_result = DB.query_single(
+        "UPDATE reviewables SET version = version + 1 WHERE id = :id AND version = :version RETURNING version",
+        version: version,
+        id: self.id
+      )
+    else
+      # We didn't supply a version to update safely, so just increase it
+      version_result = DB.query_single(
+        "UPDATE reviewables SET version = version + 1 WHERE id = :id RETURNING version",
+        id: self.id
+      )
+    end
+
+    if version_result && version_result[0]
+      self.version = version_result[0]
+    else
+      raise UpdateConflict.new
+    end
+  end
+
+end
+
+# == Schema Information
+#
+# Table name: reviewables
+#
+#  id                      :bigint(8)        not null, primary key
+#  type                    :string           not null
+#  status                  :integer          default(0), not null
+#  created_by_id           :integer          not null
+#  reviewable_by_moderator :boolean          default(FALSE), not null
+#  reviewable_by_group_id  :integer
+#  claimed_by_id           :integer
+#  category_id             :integer
+#  topic_id                :integer
+#  score                   :float            default(0.0), not null
+#  potential_spam          :boolean          default(FALSE), not null
+#  target_id               :integer
+#  target_type             :string
+#  target_created_by_id    :integer
+#  payload                 :json
+#  version                 :integer          default(0), not null
+#  latest_score            :datetime
+#  created_at              :datetime         not null
+#  updated_at              :datetime         not null
+#
+# Indexes
```

Comment:
Do we need indexes for `topic_id`, `category_id`, `created_by_id`, `reviewable_by_group_id` and `claimed_by_id`? It seems like some of this columns are used for filtering so an index is going to speed up the query once the reviewables table grows.

Reply:
@tgxworld Good question! I didn't add them because they are generally applied *after* all the other indexes that are more important. 

Almost every time you view the reviewable queue you are viewing pending (status) ordered by (score DESC). On most forums that would return less than 10 records I'd imagine unless the mods were absent, so the other filters needn't have indexes.

There might be a case where we find all the reviewables for a topic that I've forgotten, but honestly I'd prefer to deploy experimentally and watch the DB stats and add indexes as needed, rather than adding indexes to columns that might not need them.